### PR TITLE
(#3327) Improves UX of Chocolatey Install without 4.8

### LINKED ISSUE #3327: Rerunning Install Script after reboot should not be necessary
### Checklist

- [X] I have verified this is the correct repository for opening this issue.
- [X] I have verified no other issues exist related to my request.

### Is Your Feature Request Related To A Problem? Please describe.

The user-experience when installing Chocolatey on a system that lacks dotnet 4.8 is mildly annoying, and our most used flow requires:
- Running the install script
- Dotnet 4.8 being installed
- Being prompted to reboot
- Rebooting
- Running the install script again

This is less than ideal, introduces complexity, and could be improved.

### Describe The Solution. Why is it needed?

I would suggest,  given that the rest of the install seems to work fine without the reboot, that we proceed with the rest of the installation after the dotnet installation.

This would result in a system that, though needing a reboot, will come up from the reboot with a fully functional Chocolatey install - and no further required steps.

Given that Chocolatey itself prevents users from running any command beyond -v/--version when 4.8 is not installed correctly (and has an appropriately explicit error message), this should not introduce a particularly awful level of confusion.

It shouldn't break any automation, either, as the install is fairly idempotent so anyone who already is running the script, rebooting, and running it again should just work.

### Additional Context

Example of a modified install proceeding after dotnet is installed (but no reboot):
image

Example of the current error message produced when there has not been a reboot:
image

### Related Issues

_No response_

### PR DESCRIPTION
**Pull request by JPRuskin**
## Description Of Changes
This commit introduces the idea that the dotnet installation needing a reboot should not immediately halt the Chocolatey installation, but instead proceed as far as possible - then prompt for a reboot without throwing.

## Motivation and Context
The install script currently throws after dotnet 4.8 is installed, which requires the user reboot and then re-run the install script. This, at least, reduces the need for an additional step post-reboot.

## Testing
1. Build or finagle Chocolatey package with updated setup functions
    I downloaded the [Chocolatey 2.2.2 package](https://community.chocolatey.org/api/v2/package/chocolatey/2.2.2) and used 7zip to overwrite `chocolateysetup.psm1` with my changes.
2. Download [Install.ps1](https://community.chocolatey.org/install.ps1)
3. Run `Install.ps1 -ChocolateyDownloadUrl ~\Path\to\Choco.nupkg` (after execution policy / unblocking files)

### Operating Systems Testing
- Windows 2019 (without dotnet 4.8 installed)  
    image
- Windows 10 sandbox (with dotnet 4.8 available)
    image
    (no change to output)
- Windows Server 2012 (with dotnet 4.8 available, and PowerShell v3)
    image
- Windows Server 2012 (without dotnet 4.8, and PowerShell v3)
    image

## Change Types Made

* ~[ ] Bug fix (non-breaking change).~
* [x] Feature / Enhancement (non-breaking change).
* ~[ ] Breaking change (fix or feature that could cause existing functionality to change).~
* ~[ ] Documentation changes.~
* [x] PowerShell code changes.

## Change Checklist

* [?] Requires a change to the documentation.
    I am unsure if there is any reference to this behaviour in our documentation.
* [ ] Documentation has been updated.
* ~[ ] Tests to cover my changes, have been added.~
* ~[ ] All new and existing tests passed?~
* [x] PowerShell code changes: PowerShell v2 compatibility checked visually but not by running it on an actual PSv2 system

## Related Issue

Fixes #3327

### CODE REVIEW

**Comment by pauby** [comment]
> PowerShell code changes: PowerShell v2 compatibility checked?

This needs to be checked. We still support PowerShell v2.

**Comment by JPRuskin** [comment]
> > PowerShell code changes: PowerShell v2 compatibility checked?
> 
> This needs to be checked. We still support PowerShell v2.

I see nothing in my changes that is incompatible to my knowledge, but I haven't run it in PS v2.

**Comment by pauby** [comment]
> I see nothing in my changes that is incompatible, to my knowledge, but I haven't run it in PS v2.

As belt and braces, can you please run it in PS v2?

**Comment by JPRuskin** [comment]
> Can you also ensure the script says 'Chocolatey CLI' and not 'Chocolatey'?

Just to be clear, do you mean updating all instances of "Chocolatey" to "Chocolatey CLI" across the installation, or just bits that I've touched for this PR?

Examples of "bits I've touched" (in yellow), all instances that are likely to want updating if we're going to make it consistent (orange), things unlikely in my opinion to want updating (red):

image

**Comment by JPRuskin** [comment]
> > Can you also ensure the script says 'Chocolatey CLI' and not 'Chocolatey'?

@pauby Just chasing this up, as I'd like to try and get this merged before January.

**Comment by pauby** [comment]
> > > Can you also ensure the script says 'Chocolatey CLI' and not 'Chocolatey'?
> 
> @pauby Just chasing this up, as I'd like to try and get this merged before January.

For consistency it would be good to change all of the `Chocolatey` mentions that mean `Chocolatey CLI` to `Chocolatey CLI`.

**Comment by JPRuskin** [comment]
So, having an unexpected issue running this on the Windows 2012 (with PowerShell 3, no 4.8) system I have access to:
image

image

Pretty neat!

That said, it does recover after the reboot.
image

I've not seen that issue before, even on other non-4.8 having systems. Any thoughts, @gep13? We thought it would be safe to run /version, but it seems not. 4.5.2 seems to be present on this system.

I still like this change, and would prefer to not run choco.exe /version before the reboot if this is happening.

File: nuspec/chocolatey/chocolatey/tools/chocolateysetup.psm1 [hunk 1]
```diff
@@ -832,11 +839,11 @@ function Invoke-Chocolatey-Initial {
     try {
         $chocoInstallationFolder = Get-ChocolateyInstallFolder
         $chocoExe = Join-Path -Path $chocoInstallationFolder -ChildPath "choco.exe"
-        & $chocoExe -v | Out-Null
+        $runResult = & $chocoExe -v
         Write-Debug "Chocolatey execution completed successfully."
     }
     catch {
-        Write-ChocolateyWarning "Unable to run Chocolatey at this time.  It is likely that .Net Framework installation requires a system reboot"
+        Write-ChocolateyWarning "Unable to run Chocolatey at this time:`n$($runResult)"
```
Comment (gep13) [inline nuspec/chocolatey/chocolatey/tools/chocolateysetup.psm1:846]:
```suggestion
        Write-ChocolateyWarning "Unable to run Chocolatey CLI at this time:`n$($runResult)"
```

File: nuspec/chocolatey/chocolatey/tools/chocolateysetup.psm1 [hunk 2]
```diff
@@ -832,11 +839,11 @@ function Invoke-Chocolatey-Initial {
     try {
         $chocoInstallationFolder = Get-ChocolateyInstallFolder
         $chocoExe = Join-Path -Path $chocoInstallationFolder -ChildPath "choco.exe"
-        & $chocoExe -v | Out-Null
+        $runResult = & $chocoExe -v
         Write-Debug "Chocolatey execution completed successfully."
```
Comment (gep13) [inline nuspec/chocolatey/chocolatey/tools/chocolateysetup.psm1:843]:
```suggestion
        Write-Debug "Chocolatey CLI execution completed successfully."
```

**Comment by pauby** [comment]
I wanted to highlight this [issue here](https://github.com/chocolatey/choco/issues/3328), which ist not fixed by this issue, but is related.

**Comment by pauby** [comment]
> I still like this change, and would prefer to not run choco.exe /version before the reboot (on a reboot-needing run) if this is happening, rather than give up on it.

Could we detect Server 2012, put some words in a different colour (to highlight it - maybe `Write-Warning`) to say you'll get an error here, ignore it, reboot as normal and everything will be fine?

**Comment by JPRuskin** [comment]
> Could we detect Server 2012, put some words in a different colour (to highlight it - maybe Write-Warning) to say you'll get an error here, ignore it, reboot as normal and everything will be fine?

We could indeed! I'll give that a go, and we can see if we think it's enough.

**Review by gep13 — APPROVED** [review]
LGTM!

**Comment by gep13** [comment]
@JPRuskin I took the liberty of modifying your commit messages slightly (they were overrunning the standard that we use for commit messages) but other than that, this LGTM, and I will merge once the CI builds complete.

**Pull request merged**